### PR TITLE
Rustdoc copy local img

### LINKED ISSUE #32104: Include images in rustdoc output
As written by Luthaf over here

https://users.rust-lang.org/t/include-images-in-rustdoc-output/3487

"Hi rustaceans!

Images in documentation can be really useful to provide some insight on algorithms or modules organisation.

It is already possible to insert a link to an image in documentation comment
```
/// ![Alt version](url://for/this/image.png)
```
But using this method, we only get to include images already available in some place of Internet.

Is it a way to make rustdoc copy some files to the target/doc folder, so that it is easy to include specific images in the documentation? The same could go for other static files, like specific CSS or JS.

Is it possible yet? If not, do you think it is worth it?"

I this this would be awesome to include as sometimes using images is much easier when explain something than showing text.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #32104.

r? @kinnison

### CODE REVIEW

**Comment by kinnison** [comment]
@GuillaumeGomez While the approach of hashing the URL is cheap, it's also a bit risky.  Consider a library which has some image it uses from a lot of places to illustrate things.  Those places may all have subtly different paths to the same object (different numbers of `..`s for example) and so we may end up with the same potentially large resource repeated many times.  It would be better to hash the content of the file instead -- is there a reason you chose not to?

**Comment by GuillaumeGomez** [comment]
@kinnison This is a very good point. Maybe storing a checksum instead maybe and use it as name? What do you think?

**Comment by GuillaumeGomez** [comment]
I realized that we could get the full path from a path and therefore generate a same hash every time.

**Comment by kinnison** [comment]
If you've resolved the filename to a canonical path before hashing then that sounds fine to me.  It will make it harder to match exact hashes in your test though.

File: src/test/rustdoc/copy-local-img.rs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#![crate_name = "foo"]
+
+// @has static/8a40d4987fbb905
+// @has foo/struct.Enum.html
+// @has - '//img[@src="../static/8a40d4987fbb905"]' ''
```
Comment (kinnison) [inline src/test/rustdoc/copy-local-img.rs]: I fear these hashes won't be consistent between systems now that the filename is canonicalised.  Is there a cut-down test which will still work here?

**Review by kinnison — CHANGES_REQUESTED** [review]
I'm good with the approach, but I think the test will need rethinking.

File: src/test/rustdoc/copy-local-img.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/test/rustdoc/copy-local-img.rs]: Oh, didn't think about that, the hash will fail indeed. Great catch! I can add something to count files in the given folder and add a "start with" thing. Unless you have another idea?

Comment (kinnison) [reply src/test/rustdoc/copy-local-img.rs]: One option could indeed be that.  Another -- hash the file as part of the deduplication effort, but then simply number them as you add them to the `static/` dir?

**Comment by ollie27** [comment]
There will need to be some way to explicitly opt-in to this feature. As it stands this could break the docs for people already using relative paths to images. Additionally missing files should be errors.

The main issue I see with this implementation is that it resolves paths to the images relative to rustdoc's current directory. Even if that were changed to be relative to the source file, it can't be rustdoc which actually loads the file because rustdoc doesn't have access to the source files of external crates so cross-crate inlining won't work.

I think this will have to be implemented similarly to `#[doc(include)]` where the file is loaded by rustc, something like `#[doc(include_image = "../image.png")]` perhaps.

**Comment by joelpalmer** [comment]
Triaged

**Comment by Dylan-DPC** [comment]
Closing this due to inactivity.

**Pull request closed without merging**